File: frontend-offline/src/game/objects/man.ts
import { useGlobalStore } from "../../store";
import { BASE_MAN_SPEED } from "../gameConfig";
import { ManSpriteKey } from "../sprite_animations/sprite";
import { ManDirection, MapIndex, Position } from "../types";
import { BaseObj } from "./base";
import { ItemType } from "./item";
const MAX_SPEED = 12
const MAX_BOMB_NUM = 6
const MAX_BOMB_POWER = 6
export class ManObj extends BaseObj {
    // 自訂屬性
    isAlive: boolean = true
    manSpriteKey: ManSpriteKey
    dir: ManDirection = "down"
    isMoving: boolean = false
    usedBombNum: number = 0
    speed = BASE_MAN_SPEED
    bombNum = 2
    bombPower = 2
    canPassBombPosList: Position[] = []
    constructor(scene: Phaser.Scene, index: MapIndex, manSpriteKey: ManSpriteKey, userId: number) {
        super(scene, index, manSpriteKey, 5, "man");
        this.manSpriteKey = manSpriteKey
        this.sprite.setDepth(9)
    }
    setDir(dir: ManDirection) {
        if (!this.isAlive) {
            return
        }
        const animKey = `${this.manSpriteKey}_${dir}`
        this.sprite.anims.play(animKey, true)
        this.dir = dir
    }
    setMoving(v: boolean) {
        if (!this.isAlive) {
            return
        }
        this.isMoving = v
        if (!v) {
            this.sprite.anims.pause()
        } else {
            this.sprite.anims.resume()
        }
    }
    eatItem(itemType: ItemType) {

        if (itemType === 'speed') {
            this.speed = Math.min(MAX_SPEED, this.speed + 1)
        }
        else if (itemType === 'moreBomb') {
            this.bombNum = Math.min(MAX_BOMB_NUM, this.bombNum + 1)
        }
        else if (itemType === 'fire') {
            this.bombPower = Math.min(MAX_BOMB_POWER, this.bombPower + 1)
        }
    }
}

File: backend/src/game/objects/man.rs
use crate::game::config::{BASE_MAN_SPEED, MAX_BOMB_NUM, MAX_BOMB_POWER, MAX_SPEED};
use crate::game::types::{ItemType, ManDirection, ManSpriteKey, MapIndex, ObjType, Position};
use super::base::BaseObj;

pub struct ManObj {
    pub base: BaseObj,
    pub is_alive: bool,
    pub man_sprite_key: ManSpriteKey,
    pub dir: ManDirection,
    pub is_moving: bool,
    pub used_bomb_num: u32,
    pub speed: u32,
    pub bomb_num: u32,
    pub bomb_power: u32,
    pub can_pass_bomb_pos_list: Vec<Position>,
}

impl ManObj {
    pub fn new(index: &MapIndex, man_sprite_key: ManSpriteKey) -> Self {
        Self {
            base: BaseObj::new(index, ObjType::Man),
            is_alive: true,
            man_sprite_key,
            dir: ManDirection::Down,
            is_moving: false,
            used_bomb_num: 0,
            speed: BASE_MAN_SPEED,
            bomb_num: 2,
            bomb_power: 2,
            can_pass_bomb_pos_list: vec![],
        }
    }

    pub fn set_dir(&mut self, dir: ManDirection) {
        if !self.is_alive {
            return;
        }
        self.dir = dir;
    }

    pub fn set_moving(&mut self, v: bool) {
        if !self.is_alive {
            return;
        }
        self.is_moving = v;
    }

    pub fn eat_item(&mut self, item_type: &ItemType) {
        match item_type {
            ItemType::Speed => self.speed = (self.speed + 1).min(MAX_SPEED),
            ItemType::MoreBomb => self.bomb_num = (self.bomb_num + 1).min(MAX_BOMB_NUM),
            ItemType::Fire => self.bomb_power = (self.bomb_power + 1).min(MAX_BOMB_POWER),
        }
    }
}
